src/FileSystem/FileSystem.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+namespace OomphInc\WASP\FileSystem;
+
+use RuntimeException;
+
+class FileSystem implements FileSystemInterface {
+
+	use DirStackTrait;
+
+	/**
+	 * @param string [$cwd] optional starting current working directory
+	 */
+	public function __construct($cwd = null) {
+		$this->setCwd($cwd);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function resolvePath($path) {
+		// default path if null provided
+		if ($path === null) {
+			return getcwd();
+		}
+
+		// is the path relative?
+		if (!preg_match('#^/|[a-z]+://#', $path)) {
+			$path = $this->getCwd() . '/' . $path;
+		}
+
+		return is_readable($path) ? realpath($path) : $path;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getFiles($pattern, $relative = true) {
+		$cwd = $this->getCwd();
+		$files = [];
+		foreach (glob($cwd . '/' . $pattern, GLOB_BRACE) as $file) {
+			// strip off the files root path
+			$files[] = $relative ? preg_replace('#^' . preg_quote(rtrim($cwd, '/') . '/', '#') . '#', '', $file) : $file;
+		}
+		return $files;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function fileExists($path) {
+		return is_readable($this->resolvePath($path));
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function readFile($path) {
+		$file = file_get_contents($this->resolvePath($path));
+		if ($file === false) {
+			throw new RuntimeException("Could not read file '$path'");
+		}
+		return $file;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function writeFile($path, $contents) {
+		if (file_put_contents($this->resolvePath($path), $contents) === false) {
+			throw new RuntimeException("Could not write to file '$path'");
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function deleteFile($path) {
+		if ($this->fileExists($path)) {
+			if (!unlink($this->resolvePath($path))) {
+				throw new RuntimeException("Could not delete file '$path'");
+			}
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function renameFile($oldPath, $newPath) {
+		if ($this->fileExists($oldPath)) {
+			if (!rename($this->resolvePath($oldPath), $this->resolvePath($newPath))) {
+				throw new RuntimeException("Could not rename file '$oldPath'");
+			}
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function mkDir($path) {
+		if (!mkdir($this->resolvePath($path), 0777, true)) {
+			throw new RuntimeException("Could not create directory '$path'");
+		}
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function rmDir($path) {
```

Comment:
rmDir is a confusing name for this method since it deletes the contents of the directory, unlike the `rmdir` utility/PHP function

Reply:
IMO that's what every `rmdir` function _should_ do, but I see your point. Perhaps a second arg for `$force`?

Reply:
Why not just call the function `rm`, since it's (for now) literally using `rm`?

Reply:
but there's already a `deleteFile`, which falls under `rm` as well ???

Reply:
oy

Reply:
ok but do you really need a `deleteFile` too? the `rm` function i commented below would do the same thing

Reply:
'spose not. So should `$recursive` be an additional param or just the standard/non-configurable behavior?

Reply:
I don't really see the need for an additional param? I can't really think of a scenario where you would want to throw an exception if you try to delete a directory rather than a file, but maybe that's naive

Reply:
well it's just that we'd be removing the functionality of `rmDir`, which is to remove a directory only if it's empty; by a similar logic to you saying `rmDir` behaving differently from what that command normally does, the same things applies to `rm`

Reply:
that being said, I don't find `rmdir` to be very useful (hence why I coded it the way I did), and the other methods allow the user to delete a directory only if it's empty by manually check its contents first... But I'd still maybe want to name this `rmRecursive` or something like that??? But then again, the `mkdir` method forces the `-p` behavior (which again I feel should be the default), so idk

Reply:
I'm less concerned with `rmDir` matching the `rmdir` utility than the `rmdir` PHP function, so I think by that logic it's fine for this to just be called `rm`, since there's no `rm` function in PHP.

Reply:
Also, yes, I think `-r` should be a default for `rm`

Reply:
well then how do you feel about just including all of this logic in `deleteFile` for consistency among the methods as they stand now? The name doesn't suggest that it works for dirs too, technically, but neither does `fileExists`. Perhaps anything called `file` could be renamed to `path`? And to avoid drawing direct comparisons to a command/PHP function (and being more consistent, again), perhaps `mkDir` becomes `createDir`?

Reply:
or I should say, anything that is named `file` that also works on dirs

Reply:
I like it! A directory is just a special kind of file anyways (is that true on windows???)

Reply:
it's really hard to google that question because there are a million articles explaining how a file works for people who don't get computers

Reply:
lol